### ISSUE TITLE
Bug in secure comparison

### ISSUE THREAD

**Issue by GuopengLin** [secretflow/secretflow#1958]
### Issue Type

Build/Installation

### Installation Source

Binary

### Secretflow Version

Secretflow 1.13.0b0

### Operating System and Distribution

Ubuntu 22.04

### Python Version

3.10.12

### Issue Description

When performing a sequence of comparisons and arithmetic operations on secret-shared floating-point numbers, the SPU produces a logically incorrect result. A computation that should evaluate to False is incorrectly returning True.

```
# Reproduction Code
import secretflow as sf
import numpy as np
def compute(a, b):
    lt = b < a
    le = b <= a
    gt = a > b
    ge = a >= b
    eq = a == b
    neq = a != b

    nested1 = lt > ge
    nested2 = (le == eq) != (gt < neq)
    nested3 = (lt + eq) >= (ge + neq)

    return nested1 + nested2 + nested3



# ----------------- 1. 初始化 SecretFlow -----------------
# 和之前一样，在本地模拟一个包含'alice'和'bob'的隐私计算集群。
sf.init(['alice', 'bob'], address='local', log_to_driver=False)

# ----------------- 2. 定义参与方和MPC设备 -----------------
alice = sf.PYU('alice')
bob = sf.PYU('bob')

# 使用明确的配置来创建SPU设备，这是一种健壮的做法。
spu_config = {
    'nodes': [
        {'party': 'alice', 'address': '127.0.0.1:9001'},
        {'party': 'bob', 'address': '127.0.0.1:9002'},
    ],
    'runtime_config': {
        'protocol': 'SEMI2K',  # 使用标准的半诚实两方协议
        'field': 'FM64',        # 使用64位定点数计算域
    }
}
spu = sf.SPU(spu_config)

a_plain = alice(lambda: -86.4092)()
b_plain = bob(lambda: -88.2658)()
a_secret = a_plain.to(spu)
b_secret = b_plain.to(spu)

result_secret = spu(compute)(a_secret, b_secret)
result = sf.reveal(result_secret)
print(result)
```

```
# Output 
True
```

When performing the same logic directly in Python, it outputs 0 (false).

```
>>> def compute(a, b):
...     lt = b < a
...     le = b <= a
...     gt = a > b
...     ge = a >= b
...     eq = a == b
...     neq = a != b
... 
...     nested1 = lt > ge
...     nested2 = (le == eq) != (gt < neq)
...     nested3 = (lt + eq) >= (ge + neq)
... 
...     return nested1 + nested2 + nested3
...     
>>> a = -86.4092
>>> b = -88.2658
>>> compute(a, b)
0
>>>
```

**Comment by shaojian-ant** [comment]
In JAX, arithmetic operations on boolean types do not implicitly promote them to integers. An explicit cast is required, for example: ge.astype('int') + neq.astype('int').
The following JAX code snippet demonstrates a scenario where adding two boolean `True` values results in a boolean `True`.

Code:
```
import jax.numpy as jnp

a_jax = jnp.array(-86.4092)
b_jax = jnp.array(-88.2658)

ge_jax = a_jax >= b_jax
neq_jax = a_jax != b_jax

print(f"ge_jax = {ge_jax}, dtype = {ge_jax.dtype}")
print(f"neq_jax = {neq_jax}, dtype = {neq_jax.dtype}")

result3 = ge_jax + neq_jax
print(f"ge_jax + neq_jax = {result3}, dtype = {result3.dtype}")
```
Observed Output:
```
ge_jax = True, dtype = bool
neq_jax = True, dtype = bool
ge_jax + neq_jax = True, dtype = bool
```

**Comment by GuopengLin** [comment]
Thank you for the detailed explanation and for providing the JAX example. I understand that JAX's default behavior is to treat the addition of booleans as a logical AND unless they are explicitly cast to integers.

However, from the perspective of a user of this library, this behavior is unexpected and should be treated as a bug for two reasons:

1. In standard Python, booleans are a subclass of integers, and users are accustomed to them being automatically treated as 1 (True) and 0 (False) in arithmetic contexts. For example, True + True correctly evaluates to 2. Users of a Python-based library would naturally expect the + operator to perform an arithmetic sum, not a logical operation.

2. Even when working with Boolean shares , the user's intent for an operation like a + b is almost always an arithmetic addition, which would be implemented with a Boolean adder circuit. Silently performing a logical AND instead can lead to subtle and very difficult-to-debug errors in complex cryptographic protocols.

**Comment by deadlywing** [comment]
I don't believe this is a bug. 
Both NumPy and JAX NumPy follow the same design principle, which is to overload the + and * operators for boolean types. In scientific computing and data analysis, booleans are frequently used as masks, and using + and * for logical operations provides a more intuitive way to express these operations. 
For users familiar with NumPy, this convention is quite common and well-understood.

**Comment by GuopengLin** [comment]
Thank you for the response. I try it in NumPy, and it actually works the same way.
